streaming/storage/deque.go

```diff
@@ -0,0 +1,392 @@
+//Returns ErrNotFound if the list is empty
+func (dq *Deque) PopFront(value proto.Message) error {
+	//there is no need to call initialize here, since PeekFront calls initialize
+	err := dq.PeekFront(value)
+	if err != nil {
+		return err
+	}
+
+	return dq.pop(value, dq.firstFreeFrontSpot+1, dq.increaseFirstFreeFrontSpot)
+}
+
+//Removes the last element from the queue and returns it
+//Returns ErrNotFound if the list is empty
+func (dq *Deque) PopBack(value proto.Message) error {
+	//there is no need to call initialize here, since PeekBack calls initialize
+	err := dq.PeekBack(value)
+	if err != nil {
+		return err
+	}
+
+	return dq.pop(value, dq.firstFreeBackSpot-1, dq.decreaseFirstFreeBackSpot)
+}
+
+func (dq *Deque) pop(value proto.Message, index int, modifier func() error) error {
+	err := dq.peekIndex(value, index)
+	if err == ErrNotFound {
+		return ErrNotFound
+	} else if err != nil {
+		return errors.Wrap(err, "something went wrong during peek")
+	}
+
+	key := getIndexKey(index)
+
+	err = dq.tx.Delete(key)
+	if err != nil {
+		return errors.Wrap(err, "failed to delete element from queue in pop")
+	}
+
+	err = modifier()
+	if err != nil {
+		return errors.Wrap(err, "failed to actualize queue attributes in pop")
+	}
+
+	return nil
+}
+
+//Returns the first element of the queue without removing it
+//Returns ErrNotFound if the queue is empty
+func (dq *Deque) PeekFront(value proto.Message) error {
+	if !dq.initializedFront {
+		err := dq.initializeFront()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in queue.PeekFront")
+		}
+	}
+
+	return dq.peekIndex(value, dq.firstFreeFrontSpot+1)
+}
+
+//Returns the last element of the queue without removing it
+//Returns ErrNotFound if the queue is empty
+func (dq *Deque) PeekBack(value proto.Message) error {
+	if !dq.initializedBack {
+		err := dq.initializeBack()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in queue.PeekBack")
+		}
+	}
+
+	return dq.peekIndex(value, dq.firstFreeBackSpot-1)
+}
+
+//we don't check if the queue has been initialized here, since we only
+//call peekIndex from PeekFront/PeekBack, which both initialize the queue
+func (dq *Deque) peekIndex(value proto.Message, index int) error {
+	key := getIndexKey(index)
+
+	data, err := dq.tx.Get(key)
+	if err == ErrNotFound {
+		return ErrNotFound
+	} else if err != nil {
+		return errors.Wrap(err, "an error occurred during Get")
+	}
+
+	err = proto.Unmarshal(data, value)
+	if err != nil {
+		return errors.Wrap(err, "couldn't unmarshal the element in peek")
+	}
+
+	return nil
+}
+
+//Clears all contents of the queue including the metadata
+func (dq *Deque) Clear() error {
+	for dq.firstFreeBackSpot-dq.firstFreeFrontSpot > 1 {
+		key := getIndexKey(dq.firstFreeFrontSpot + 1)
+		err := dq.tx.Delete(key)
+
+		if err != nil {
+			return errors.Wrap(err, "couldn't clear the list")
+		}
+
+		//we call increaseFirstFree instead of simply dq.firstFree += 1 because
```

Comment:
If a machine dies during clear, then this structure will be created anew and the whole transaction won't have happened.